2021/day15/day15.java

```diff
@@ -0,0 +1,146 @@
+                    if (tmpRisk > 9)
+                        tmpRisk = tmpRisk % 10 + 1;
+                    newLine.add(tmpRisk);
+                }
+                mapOfRisks.add(newLine);
+            }
+        }
+        return mapOfRisks;
+    }
+
+    /*
+     * Implementation of Dijkstra's algorithm for finding shortest path between
+     * two nodes. It goes like this:
+     * Do the following as long as we did not reach destination:
+     * 1. Visit all neighbour nodes (that are not in state "already visited")
+     * of the current node. Check their distance (risk) from the starting
+     * position and add them to priority queue. Set their position to
+     * "already visited".
+     * 2. From the priority queue take node that is the closest to starting
+     * position and move there.
+     * 3. Update current position and distance (risk).
+     */
+    public static int calculateRisk(boolean isPart2) {
+        ArrayList<ArrayList<Integer>> mapOfRisk = readInput("data.txt");
+        if (isPart2) {
+            mapOfRisk = modifyInput(mapOfRisk, 5);
+        }
+        ArrayList<ArrayList<Boolean>> isVisited = new ArrayList<>();
+        for (ArrayList<Integer> element : mapOfRisk) {
+            ArrayList<Boolean> line = new ArrayList<>
+                                (Collections.nCopies(mapOfRisk.size(), false));
+            isVisited.add(line);
+        }
+        PriorityQueue<Position> queOfPositions = new PriorityQueue<>();
+        int x = 0;
+        int y = 0;
+        int risk = 0;
+        while (x != mapOfRisk.size() - 1 || y != mapOfRisk.size() - 1) {
+            // visit right neighbour
+            if (x < mapOfRisk.size() - 1 && !isVisited.get(y).get(x + 1)) {
+                Position position = new Position(x + 1, y,
+                        risk + mapOfRisk.get(y).get(x + 1));
+                queOfPositions.add(position);
+                isVisited.get(y).set(x + 1, true);
+            }
+            // visit left neighbour
+            if (x > 0 && !isVisited.get(y).get(x - 1)) {
+                Position position = new Position(x - 1, y,
+                        risk + mapOfRisk.get(y).get(x - 1));
+                queOfPositions.add(position);
+                isVisited.get(y).set(x - 1, true);
+            }
+            // visit top neighbour
+            if (y > 0 && !isVisited.get(y - 1).get(x)) {
+                Position position = new Position(x, y - 1,
+                        risk + mapOfRisk.get(y - 1).get(x));
+                queOfPositions.add(position);
+                isVisited.get(y - 1).set(x, true);
+            }
+            // visit bottom neigbour
+            if (y < mapOfRisk.size() - 1 && !isVisited.get(y + 1).get(x)) {
+                Position position = new Position(x, y + 1,
+                        risk + mapOfRisk.get(y + 1).get(x));
+                queOfPositions.add(position);
+                isVisited.get(y + 1).set(x, true);
+            }
```

Comment:
another place where a `neighbours` method is the way, there is too much repetition in these four clauses. Having such a function, you just iterate through the neighbours, check if they were already visited and if not add them to the queue